Fixing Imgur .gif links by replacing the extension with .gifv

spiral6 opened the pull request:
This won't affect any link except the ones that are .gifs. Because Lapis links to the direct image every time, and not the imgur page, this should work for every instance of a .gif.

spiral6 committed:
is it really this easy...

spiral6 commented:
Resolves #30

kupiakos commented:
How's it tested? IIRC, Imgur will sometimes say an image is a `.jpg` even if it is actually a GIF.

Also, you can put `Resolves #30` in your Pull Request or commit messages. That way the issue will automatically be referenced, and if I pull it in, the issue will automatically be closed.

Personally I'd prefer for commits to match the [standard style](https://git-scm.com/book/ch5-2.html) in the future.

spiral6 commented:
I personally haven't experienced a single instance of when imgur has returned a URL for a .gif incorrectly after uploading. That being said, I can implement a check to see if the URL returns a gif or differently, and implement logic to replace as needed.

spiral6 committed:
Resolves #30

Added mime-type check to see if image is gif or not, and if so, replace URL's extension

spiral6 commented on plugins/imgur.py:
```diff
@@ -128,7 +128,11 @@ def export_submission(self,
             if is_album:
                 results['link_display'] = '[Imgur Album](https://imgur.com/a/%s)  \n' % album['id']
             else:
-                results['link_display'] = '[Imgur](%s)  \n' % images[0]['link'].replace('http', 'https').replace('gif, 'gifv')
```
there's a typo in here, but this line is no longer in the commit anyway

kupiakos commented on plugins/imgur.py:
```diff
@@ -128,7 +128,11 @@ def export_submission(self,
             else:
-                results['link_display'] = '[Imgur](%s)  \n' % images[0]['link'].replace('http', 'https')
+                picture_url = images[0]['link'].replace('http', 'https')
+                with urllib.request.urlopen(picture_url) as response:
```
Use [`requests.head`](https://github.com/kupiakos/LapisMirror/blob/70676ac8745ff2a06970798356f0cb8b3b960251/plugins/puush.py#L73) to get the MIME type and compare against that.

spiral6 committed:
Added mime-parsing and regex substitution of URL extension

kupiakos commented on plugins/imgur.py:
```diff
@@ -128,7 +129,15 @@ def export_submission(self,
+                mime_text = r.headers.get('Content-Type')
+                mime = mimeparse.parse_mime_type(mime_text)
+                if mime[1] == 'gif':
+                    picture_url = re.sub('\.gif|jpg|png', '.gifv', picture_url)
```
Why not just do a `rsplit` and replace there? More importantly, though, `'\.` is an invalid escape sequence and will be deprecated. Either use a raw string or double-backslash.

spiral6 replied:
Sorry, Python isn't my first language. I'm not very used to it. So I didn't know that `\.` was deprecated. I'm not quite sure what you mean by a `rsplit`, what would I be splitting by? 

Does this example make any sense?
```
pictureurl = 'https://i.imgur.com/iGAq9SV.gif'
split = pictureurl.rsplit('.')
split[3] = 'gifv'
pictureurl = '.'.join(split)
```

This would forgo the use of regex entirely, anyway.

kupiakos replied:
Hrm, looking at it a little more closely, it may be safer to use a regex than a split, but it wouldn't be this one exactly. The way it's constructed right now, this could happen:

```python
>>> picture_url = 'https://i.imgur.com/ijpg9SV.gif'
>>> re.sub(r'\.gif|jpg|png', '.gifv', picture_url)
'https://i.imgur.com/i.gifv9SV.gifv'
```

Which is obviously not what is wanted.
I'd probably do something more like this:

```python
>>> re.sub(r'(\.\w+)?$', '.gifv', 'https://i.imgur.com/ijpg9SV.gif')
'https://i.imgur.com/ijpg9SV.gifv'
>>> re.sub(r'(\.\w+)?$', '.gifv', 'https://i.imgur.com/ijpg9SV.jpg')
'https://i.imgur.com/ijpg9SV.gifv'
>>> re.sub(r'(\.\w+)?$', '.gifv', 'https://i.imgur.com/ijpg9SV')
'https://i.imgur.com/ijpg9SV.gifv'
```

I use the `r'text \ here'` syntax which allows me to insert backslashes in a string without having to escape them. Breaking down the regular expression `(\.\w+)?$`, it's:

- `(...)?$`: Optionally select a group at the end of the string
- `\.`: Match the literal dot character (I can use one `\` because I used `r''` for the whole string)
- `\w`: Match a word character (alphabetical, number, or underscore)
- `+`: Match one or more of the previous expression (in this case, `\w`)

So, all together, it's "If it exists, select the bit at the end starting with a dot and a bunch of word characters. Otherwise, just select the end. Then, replace the selection with `.gifv`."

Take a look at the [re](https://docs.python.org/3/library/re.html) module for more info.

spiral6 replied:
Alright, that should do it then. Thanks a lot. I was going to refine the regular expression, but you did it before I could. And thanks for the explanation on raw strings. I know C# and Python have them, but I'm just not used to them as I'm mainly a Java programmer.

spiral6 committed:
Resolves #30

Thanks @kupiakos, that should do the trick.

kupiakos committed:
Adjust spacing for .gifv conversion
